common/k8s/src/event_source.rs

```diff
@@ -0,0 +1,408 @@
+                                                                .deletion_timestamp
+                                                                .map(|t| {
+                                                                    info!("Ignoring events before {}", t.0 - chrono::Duration::seconds(2));
+                                                                    NonZeroI64::new(t.0.timestamp() - 2)
+                                                                })
+                                                                .flatten(),
+                                                        );
+                                                        Cont::Break
+                                                    }
+                                                    _ => Cont::Cont,
+                                                }
+                                            }
+                                        })
+                                        .filter_map(|e| async move {
+                                            match e {
+                                                Cont::Cont => None,
+                                                Cont::Break => Some(((), ())),
+                                            }
+                                        });
+                                    pin_mut!(stream);
+                                    stream.next().await
+                                }
+                            } else {
+                                Some(((), ()))
+                            }
+                        }
+                    };
+                    //pin_mut!(waiter);
+                    let waiting_stream = unfold((), waiter).map(|_| Ok(StreamElem::Waiting));
+
+                    let events: Api<Event> = Api::all(a_client.as_ref().clone());
+                    let params = ListParams::default();
+
+                    let event_stream = try_flatten_touched(watcher(events, params))
+                        .map_err(K8sEventStreamError::WatcherError)
+                        .filter({
+                            move |ref event| {
+                                let earliest = previous_event_logger_delete_time.clone();
+                                let ret = earliest.load().and_then(|earliest| {
+                                    let earliest =
+                                        chrono::NaiveDateTime::from_timestamp(earliest.into(), 0);
+                                    event.as_ref().ok().and_then(|e| {
+                                        e.last_timestamp
+                                            .as_ref()
+                                            .map(|l| earliest < l.0.naive_utc())
+                                    })
+                                });
+                                async move { ret.unwrap_or(true) }
+                            }
+                        })
```

Comment:
If I'm understanding correctly, once the `event_stream` is unblocked by the `waiting_stream`, it's going to use the shared resource `previous_event_logger_delete_time` to filter out all the old events. This approach is a bit janky as it requires potentially reading through a long list of events to catch up. I'm not sure how many events Kube keeps around and for how long so it might not be a real cause concern.

If the Kube API allows us to specify a start time, maybe we can use your `RestartingStream` to build a fresh new stream once the `waiting_stream` unblocks. It does require a bit of rethinking tho... and might possible be even more confusing...

Reply:
Unfortunately you can't do > or < on labels with the kube api, so we can't set a start point there, I tried that first and wasn't able to find any mechanism to allow it.

By default kubernetes expires events after 1 hour has passed and pages them back 500 at a time so even completely reprocessing a stream of a few hundred thouasnd events shouldn't be a _huge_ load. And once it's caught up it's there isn't much work to do. I'd definitely like to do some testing around this though. Unfortunately k8s doesn't even necessarily ship events in order, so we have to keep the filter in place rather that just skipping until we're at desired point... https://github.com/kubernetes/kubernetes/issues/29838